setuptools/tests/test_build_meta.py

```diff
@@ -981,3 +982,99 @@ def test_system_exit_in_setuppy(monkeypatch, tmp_path):
     with pytest.raises(SystemExit, match="some error"):
         backend = BuildBackend(backend_name="setuptools.build_meta")
         backend.get_requires_for_build_wheel()
+
+
+class TestTypeInformationIncludedByDefault:
+    dont_include_package_data = """
+    [project]
+    name = "foo"
+    version = "1"
+    [tools.setuptools]
+    include-package-data = false
+    """
+
+    exclude_type_info = """
+    [tool.setuptools.exclude-package-data]
+    "*" = ["py.typed", "*.pyi"]
+    """
+
+    package_1 = {
+        "foo": {
+            "bar.pyi": "",
+            "py.typed": "",
+        }
+    }
+
+    package_2 = {
+        "foo": {
+            "bar": {
+                "py.typed": "",
+                "mod.pyi": "",
+            }
+        }
+    }
+
+    package_3 = {
+        "foo": {
+            "namespace": {
+                "foo.pyi": "",
+            },
+            "__init__.pyi": "",
+            "py.typed": "",
+        }
+    }
+
+    packages_to_test = [package_1, package_2, package_3]
+
+    def is_type_information_file(self, filename):
+        basename = os.path.basename(filename)
+        return basename.endswith(".pyi") or basename == "py.typed"
+
+    def get_type_files(self, file_spec):
+        output = set()
+        for key in file_spec.keys():
+            if isinstance(file_spec[key], str):
+                if self.is_type_information_file(key):
+                    output.add(key)
+            else:
+                output.update(
+                    key + "/" + f for f in self.get_type_files(file_spec[key])
+                )
+        return output
+
+    @pytest.fixture(params=itertools.product(packages_to_test, [True, False]))
+    def file_spec_and_expected(self, request):
+        file_spec, exclude_type_information = request.param
+        pyproject = self.dont_include_package_data
+        if exclude_type_information:
+            pyproject += self.exclude_type_info
+        file_spec["pyproject.toml"] = pyproject
+
+        if exclude_type_information:
+            expected = set()
+        else:
+            expected = self.get_type_files(file_spec)
+
+        yield file_spec, expected
+
+    def test_type_information_always_included(
+        self, monkeypatch, tmp_path, file_spec_and_expected
+    ):
+        """Setuptools should include type information in the wheel (py.typed, *.pyi)."""
+        file_spec, expected = file_spec_and_expected
```

Comment:
Thank you very much for working so thoroughly on the tests.
You came up with some very clever fixture definition 😝.

I wonder however if it would be better to have less automation in this test and more explicit expectations, just for the sake of being easier to reason about (for the test reader).

For example, you could define explicitly both the directory structure and the expected type files, something like the following:


```python
EXAMPLES = {
   "single_shallow_package": {
       "directory_structure": {
            "foo": {
                "bar.pyi": "",
                "py.typed": "",
            },
       }
       "type_files": ["foo/bar.pyi", "foo/py.typed"],
   },
   "nested_inside_namespace": {...},
   "namespace_nested_inside_regular": {...},
   ...
}

CONFIG = {
   "default": "...",
   "dont_include_package_data": "...",
   "exclude_package_data": "...",   
}
```

This way you can parametrize on the dict keys and it would be easier to figure out what is being asserted, for example:[^1]

[^1]: I like to use `dict.keys()` in Pytest fixtures because it makes the test report nicer, but please feel free to use whatever approach you prefer.

```python
@pytest.mark.parametrize("config", ["default", "dont_include_package_data"])
@pytest.mark.parametrize("example", EXAMPLES.keys())
def test_type_files_included_by_default(self, example):
    structure = self.EXAMPLE[example]["directory_structure"]
    type_files = self.EXAMPLE[example]["type_files"]
    ...
    assert type_files in wheel_contents

@pytest.mark.parametrize("config", ["exclude_package_data"]))
@pytest.mark.parametrize("example", EXAMPLES.keys())
def test_type_files_can_be_excluded(self, example):
    ...
    assert type_files not in wheel_contents
```

Reply:
I rewrote the tests now into test_build_py. There are 10 tests in total though because of the parametrisation. Should some of these be removed? Also, I wasn't sure how to address the fact that the tests have a lot of duplicated code (and I wasn't sure whether this should be considered an issue). Previously I was using a slightly complicated fixture as you mentioned, and I agree it made it hard to understand the tests. Perhaps putting the tests in a class would help (but this could also make the tests harder to understand).

Reply:
I gave it a try to refactor some of it... not the best job in the universe, but I hope it is enough...

I also removed `build_py.run()` to make them fast (to minimise the impact of having 10 new tests). If I am not wrong, `build_py.get_outputs()` don't need `.run()` only `.finalize_options()`